src/Bicep.Core/TypeSystem/DeployTimeConstantVisitor.cs

```diff
@@ -53,47 +51,71 @@ public static void ValidateDeployTimeConstants(SemanticModel model, IDiagnosticW
+            // which causes state to be left over after processing a peer declaration
+            // let's clean it up
+            this.bodyType = null;
+            ResetState();
+
             // Once https://github.com/Azure/bicep/issues/1177 is fixed,
             // it should be possible to use 
             // model.GetSymbolInfo(syntax.Body) is ObectType
             // Currently propertyFlags are not propagated
-            if (model.GetSymbolInfo(syntax) is ResourceSymbol resourceSymbol &&
-                resourceSymbol.Type is ResourceType resourceType &&
-                resourceType.Body is ObjectType bodyObj)
+            var symbol = model.GetSymbolInfo(syntax);
+            switch(symbol)
             {
-                this.bodyObj = bodyObj;
+                case ResourceSymbol { IsCollection: false, Type: ResourceType { Body: ObjectType bodyObj } }:
+                    this.bodyType = bodyObj;
+                    break;
+
+                case ResourceSymbol { IsCollection: true, Type: ArrayType { Item: ResourceType { Body: ObjectType bodyObj } } }:
+                    this.bodyType = bodyObj;
+                    break;
             }
+
             base.VisitResourceDeclarationSyntax(syntax);
-            this.bodyObj = null;
+            this.bodyType = null;
         }
 
         public override void VisitModuleDeclarationSyntax(ModuleDeclarationSyntax syntax)
         {
+            // in certain cases, errors will cause this visitor to short-circuit, 
+            // which causes state to be left over after processing a peer declaration
+            // let's clean it up
+            this.bodyType = null;
+            ResetState();
+
             // Once https://github.com/Azure/bicep/issues/1177 is fixed,
             // it should be possible to use 
             // model.GetSymbolInfo(syntax.Body) is ObectType
             // Currently propertyFlags are not propagated
-            if (model.GetSymbolInfo(syntax) is ModuleSymbol moduleSymbol &&
-                moduleSymbol.Type is ModuleType moduleType &&
-                moduleType.Body is ObjectType bodyObj)
+            var symbol = model.GetSymbolInfo(syntax);
+            switch(symbol)
             {
-                this.bodyObj = bodyObj;
+                case ModuleSymbol { IsCollection: false, Type: ModuleType { Body: ObjectType bodyType } }:
+                    this.bodyType = bodyType;
+                    break;
+
+                case ModuleSymbol { IsCollection: true, Type: ArrayType { Item: ModuleType { Body: ObjectType bodyType } } }:
+                    this.bodyType = bodyType;
+                    break;
             }
+
             base.VisitModuleDeclarationSyntax(syntax);
-            this.bodyObj = null;
+            this.bodyType = null;
         }
         #endregion
 
         public override void VisitObjectSyntax(ObjectSyntax syntax)
         {
-            if (this.bodyObj == null)
+            if (this.bodyType == null)
```

Comment:
>bodyType [](start = 21, length = 8)

Renamed because "Obj" was making me thing it was `ObjectSyntax` but we are dealing with `ObjectType`.